_docs/how-to/python-package.md

```diff
@@ -0,0 +1,207 @@
+but proper versioning support is critical when sharing your library more widely,
+which will require promoting your package to PyPI.
+Tools that scan for security vulnerabilities may also struggle with such dependencies.
+
+You’ll need to include git in your charm’s build dependencies to use a GitHub-hosted library in your charm:
+
+```yaml
+parts:
+  charm:
+    build-packages: [git]
+```
+
+Then you can specify the dependency in your requirements like this:
+
+```
+ops @ git+https://github.com/canonical/operator@main
+```
+
+If you don’t specify a branch or tag, it will default to `main`.
+This is probably sufficient for early prototyping and internal use.
+If you push git tags for releases or do releases on GitHub, you could point to the exact release tag, e.g. `@v2.18.1`.
+Another approach which could be useful for prototyping would be to pin on a branch,
+e.g. `stable/candidate/beta/edge`, `dev/main`, `v0/v1/v2`, etc.
+These approaches may be useful as your library stabilises internally, but if you find yourself thinking about a scheme like this,
+it’s probably a sign to switch to PyPI and use semantic versioning.
+
+If your package is in a subdirectory of your repository
+(for example if you use a monorepo,
+or collect your libraries into a single repository
+-- or if you’re just developing in an existing repository while prototyping):
+
+```
+ops-testing @ git+https://github.com/canonical/operator@main#subdirectory=testing
+```
+
+If you're using `pyproject.toml` (recommended!):
+
+```toml
+[project]
+dependencies = [
+  "ops-testing @ git+https://github.com/canonical/operator@main#subdirectory=testing",
+]
+```
+
+For `poetry` see [here](https://python-poetry.org/docs/dependency-specification/#git-dependencies).
+
+(python-package-distribution-local)=
+### Local Files
+
+If you're developing a Python package in the same repository as your charm(s),
+it may be simplest to skip distribution and use the local files when packing the charm.
+This way, assuming a freshly checked out copy of the repository,
+the git commit fully specifies the contents of both your charm code and the package.
+For example, if you had the following structure:
+
+```
+$repo/
+    $charm/
+        src/charm.py
+        pyproject.toml
+    $package/
+        src/charmlibs/$libname/__init__.py
+        pyproject.toml
+```
+
+Then you could leave `$package` out of your `$charm/pyproject.toml` during development.
+
+To provision a development virtual environment in `$repo` you could `uv venv` and then `uv pip install -e ./$charm -e ./$package`.
+Using editable installs ensures that the virtual environment reflects all changes made to either `$charm` or `$package`.
+To provision a development virtual environment in `$charm` you could `uv sync` and then `uv pip install -e ../$package`.
+Since `$package` doesn't depend on `$charm`, its development virtual environment doesn't require any special commands, just a `uv sync`.
```

Comment:
I think these would be better as a couple of multi-line bash blocks like you have below. Describe what you're going to do, "you could use what's called an [editable install](...):" and then present the couple of lines. It gets lost a bit in the paragraph format.